R/dtc_create_iso8601.R

```diff
@@ -0,0 +1,415 @@
+#'   meaning to check against a selection of validated formats in
+#'   [dtc_formats][sdtm.oak::dtc_formats]; or to have a more permissible
+#'   interpretation of the formats.
+#'
+#' @examples
+#' # Converting dates
+#' create_iso8601(c("2020-01-01", "20200102"), .format = "y-m-d")
+#' create_iso8601(c("2020-01-01", "20200102"), .format = "ymd")
+#' create_iso8601(c("2020-01-01", "20200102"), .format = list(c("y-m-d", "ymd")))
+#'
+#' # Two-digit years are supported
+#' create_iso8601(c("20-01-01", "200101"), .format = list(c("y-m-d", "ymd")))
+#'
+#' # `.cutoff_2000` sets the cutoff for two-digit to four-digit year conversion
+#' # Default is at 68.
+#' create_iso8601(c("67-01-01", "68-01-01", "69-01-01"), .format = "y-m-d")
+#'
+#' # Change it to 80.
+#' create_iso8601(c("79-01-01", "80-01-01", "81-01-01"), .format = "y-m-d", .cutoff_2000 = 80)
+#'
+#' # Converting times
+#' create_iso8601("15:10", .format = "HH:MM", .check_format = FALSE)
+#' create_iso8601("2:10", .format = "HH:MM", .check_format = FALSE)
+#' create_iso8601("2:1", .format = "HH:MM", .check_format = FALSE)
+#' create_iso8601("02:01:56", .format = "HH:MM:SS", .check_format = FALSE)
+#' create_iso8601("020156.5", .format = "HHMMSS", .check_format = FALSE)
+#'
+#' # Converting date-times
+#' create_iso8601("12 NOV 202015:15", .format = "dd mmm yyyyHH:MM", .check_format = FALSE)
+#'
+#' # Indicate allowed missing values to make the parsing pass
+#' create_iso8601("U DEC 201914:00", .format = "dd mmm yyyyHH:MM", .check_format = FALSE)
+#' create_iso8601("U DEC 201914:00", .format = "dd mmm yyyyHH:MM", .check_format = FALSE, .na = "U")
+#'
+#' create_iso8601("NOV 2020", .format = "m y", .check_format = FALSE)
+#' create_iso8601(c("MAR 2019", "MaR 2020", "mar 2021"), .format = "m y", .check_format = FALSE)
+#'
+#' create_iso8601("2019-04-041045-", .format = "yyyy-mm-ddHHMM-", .check_format = FALSE)
+#'
+#' create_iso8601("20200507null", .format = "ymd(HH:MM:SS)", .check_format = FALSE)
+#' create_iso8601("20200507null", .format = "ymd((HH:MM:SS)|null)", .check_format = FALSE)
+#'
+#' # Fractional seconds
+#' create_iso8601("2019-120602:20:13.1230001", .format = "y-mdH:M:S", .check_format = FALSE)
+#' @export
+create_iso8601 <- function(..., .format, .fmt_c = fmt_cmp(), .na = NULL, .cutoff_2000 = 68L, .check_format = FALSE) {
+
+  assert_fmt_c(.fmt_c)
+
+  dots <- rlang::dots_list(...)
+
+  if (rlang::is_empty(dots)) {
+    return(character())
+  }
+
+  # Check if all vectors in `dots` are of character type.
+  if (!identical(unique(sapply(dots, typeof)), "character")) {
+    rlang::abort("All vectors in `...` must be of type character.")
+  }
+
+  # Check if all vectors in `dots` are of the same length.
+  n <- unique(lengths(dots))
+  if (!identical(length(n), 1L)) {
+    rlang::abort("All vectors in `...` must be of the same length.")
+  }
+
+  if (!identical(length(dots), length(.format))) {
+    rlang::abort("Number of vectors in `...` should match length of `.format`.")
+  }
```

Comment:
This doesn't work in debug mode when I give 

dots <- c("UN UNK 201914:00", "UN JAN 2021")
.format = "dd mmm yyyy"

Reply:
You would have to do it like this:

``` r
library(sdtm.oak)
datetimes <- c("UN UNK 201914:00", "UN JAN 2021")
create_iso8601(datetimes, .format = list(c("dd mmm yyyy", "dd mmm yyyyHH:MM")), .na = c("UN", "UNK"))
#> [1] "2019----T14:00" "2021-01"
```

<sup>Created on 2023-11-22 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Example added: https://github.com/pharmaverse/sdtm.oak/pull/21/commits/81336b18f5fdd41acc93c1a657bd0ea95ad765d8.